prometheus-metrics-model/src/test/java/io/prometheus/metrics/model/snapshots/PrometheusNamingTest.java

```diff
@@ -96,4 +104,435 @@ public void testEmptyUnitName() {
     assertThatExceptionOfType(IllegalArgumentException.class)
         .isThrownBy(() -> sanitizeUnitName(""));
   }
+
+  @Test
+  public void testMetricNameIsValid() {
+    assertThat(validateMetricName("Avalid_23name")).isNull();
+    assertThat(validateMetricName("_Avalid_23name")).isNull();
+    assertThat(validateMetricName("1valid_23name")).isEqualTo("The metric name contains unsupported characters");
+    assertThat(validateMetricName("avalid_23name")).isNull();
+    assertThat(validateMetricName("Ava:lid_23name")).isNull();
+    assertThat(validateMetricName("a lid_23name")).isEqualTo("The metric name contains unsupported characters");
+    assertThat(validateMetricName(":leading_colon")).isNull();
+    assertThat(validateMetricName("colon:in:the:middle")).isNull();
+    assertThat(validateMetricName("")).isEqualTo("The metric name contains unsupported characters");
+    assertThat(validateMetricName("a\ud800z")).isEqualTo("The metric name contains unsupported characters");
+  }
+
+  @Test
+  public void testLabelNameIsValid() {
+    try (MockedStatic<PrometheusNaming> mock = mockStatic(PrometheusNaming.class, CALLS_REAL_METHODS)) {
+      // Mock the validation scheme to use UTF-8 validation for this test
+      mock.when(PrometheusNaming::getValidationScheme)
+          .thenReturn(ValidationScheme.UTF_8_VALIDATION);
+      
+      // These assertions now use UTF-8 validation behavior
+      assertThat(isValidLabelName("Avalid_23name")).isTrue();
+      assertThat(isValidLabelName("_Avalid_23name")).isTrue();
+      assertThat(isValidLabelName("1valid_23name")).isTrue();
+      assertThat(isValidLabelName("avalid_23name")).isTrue();
+      assertThat(isValidLabelName("Ava:lid_23name")).isTrue();
+      assertThat(isValidLabelName("a lid_23name")).isTrue();
+      assertThat(isValidLabelName(":leading_colon")).isTrue();
+      assertThat(isValidLabelName("colon:in:the:middle")).isTrue();
+      assertThat(isValidLabelName("a\ud800z")).isFalse();
+    }
+    
+    assertThat(isValidLabelName("Avalid_23name")).isTrue();
+    assertThat(isValidLabelName("_Avalid_23name")).isTrue();
+    assertThat(isValidLabelName("1valid_23name")).isFalse();
+    assertThat(isValidLabelName("avalid_23name")).isTrue();
+    assertThat(isValidLabelName("Ava:lid_23name")).isFalse();
+    assertThat(isValidLabelName("a lid_23name")).isFalse();
+    assertThat(isValidLabelName(":leading_colon")).isFalse();
+    assertThat(isValidLabelName("colon:in:the:middle")).isFalse();
+    assertThat(isValidLabelName("a\ud800z")).isFalse();
+  }
+
+  @Test
+  public void testEscapeName() {
+    // empty string
+    String got = escapeName("", EscapingScheme.UNDERSCORE_ESCAPING);
```

Comment:
avoid reassigning variables - I think the expression for `got` can be inlined directly